tools/lintapidiff.ml

```diff
@@ -0,0 +1,268 @@
+            Format.eprintf "%a@." Location.report_error;
+            Error `Exit)
+      (fun () -> Misc.remove_file tmp)
+end
+
+module Diff = struct
+  type seen_info = {
+    not_seen: version option;
+    first_seen: version;
+    deprecated: bool;
+  }
+
+  let err k (loc, msg, a, b) =
+    let info_seen ppf = function
+      | None ->
+          Format.fprintf ppf "%s was not seen in any analyzed version" k
+      | Some a ->
+          begin match a.not_seen with
+          | Some v -> Format.fprintf ppf "%s was not seen in version %s" k v
+          | None -> Format.fprintf ppf "%s was seen in all analyzed versions" k
+          end;
+          Format.fprintf ppf "@,%s was seen in version %s" k a.first_seen;
+          if a.deprecated then
+            Format.fprintf ppf "@,%s was marked as deprecated" k
+    in
+    let info_latest ppf = function
+      | None -> Format.fprintf ppf "%s was deleted in HEAD" k
+      | Some s ->
+          begin match s.Doc.since with
+          | Some v -> Format.fprintf ppf "%s has @since %s" k v
+          | None -> Format.fprintf ppf "%s has no @since annotation" k
+          end;
+          if s.Doc.deprecated then
+            Format.fprintf ppf "@,%s is marked as deprecated" k
+    in
+    Location.errorf ~loc "@[%s %s@,%a@,%a@]" msg k info_seen a info_latest b |>
+    Format.eprintf "%a@." Location.report_error
+
+  let parse_file_at_rev ~path (prev,accum) rev =
+    let merge _ a b = match a, b with
+      | Some a, Some b ->
+          Some { a with  deprecated=b.deprecated }
+      | None, Some a -> Some { a with not_seen=prev }
+      | Some _, None -> None (* deleted *)
+      | None, None -> assert false
+    in
+    let empty = {not_seen=None;first_seen="";deprecated=false} in
+    let f = Ast.parse_file ~orig:path ~init:empty ~f:(fun _ _ attrs ->
+        { not_seen=None;first_seen=rev; deprecated=Doc.is_deprecated attrs }) in
+    let map = match Git.with_show ~f rev path with
+      | Ok r -> r
+      | Error `Not_found -> IdMap.empty
+      | Error `Exit -> raise Exit in
+    Some rev, IdMap.merge merge accum map
+
+  let version_of_string str =
+    try Scanf.sscanf str "%u.%u.%u" (fun a b c -> (a,b,c))
+    with _ -> Scanf.sscanf str "%u.%u" (fun a b -> (a,b,0))
+
+  let same_version v1 v2 =
+    version_of_string v1 = version_of_string v2
+
+  let check_changes ~first ~last default k seen latest =
+    let is_old v =
+      version_of_string v < (4,0,0) || same_version v first
+    in
+    if k = "type Pervasives.format6" then None (* was built-in before *)
+    else let open Doc in
+    match seen, latest with
+    | None, None -> assert false
+    | _, Some {has_doc_parent=false;has_doc=false;_} -> None (* undocumented *)
+    | Some {deprecated=true;first_seen;_}, None -> None (* deleted deprecated *)
+    | Some _, None ->
+        Some (default, "deleted non-deprecated", seen, latest)
+    | _, Some {deprecated=true;since=None;_} -> None (* marked as deprecated *)
```

Comment:
I think that if something is marked as deprecated users would be better of avoiding to use it, than figuring out the required compiler version constraints where they can use it. e.g. `Array.make_float` was rather short-lived: introduced in 4.02.0, then marked as deprecated, removing the since annotation.